### PR TITLE
Support weekdays in parse_datetime

### PR DESCRIPTION
**Pull request by philolo1**
@tertsdiepraam @cakebaker @sylvestre 
This commit resolves issue #23.

Adds parse_weekday function that uses chrono weekday parser with a map for edge cases. Adds tests cases to make sure it works correctly.

### CODE REVIEW

**Comment by philolo1** [comment]
@tertsdiepraam can you link to that pr?

**Comment by philolo1** [comment]
So should i wait for https://github.com/uutils/parse_datetime/pull/25 to be merged and then refactor?

**Comment by tertsdiepraam** [comment]
Hmm yeah I'm not really sure what the best course of action is. The situation is a bit unfortunate. I guess it depends on a few things:

- Will #25 be ready soon?
- How relevant is this PR to #25, i.e. can we reuse significant parts of this with `nom`?

The first question is kinda tough. I think a basic `nom` version of this library _should_ be merged soon, just to get the transition going. Which might not be the entirety of #25. The second question is interesting. At the least the tests are relevant after the introduction of `nom`.

Maybe another approach is to transition to `nom` from the bottom up. What I mean with that is that we port small parts first and gradually build up to parsing the whole thing with `nom`. For example, if we make a `nom`-based` function to parse weekdays, we can easily integrate that later.

I'm really just explaining my thoughts here. I'm not yet really sure what the best approach is. What do you think?

**Comment by philolo1** [comment]
> Hmm yeah I'm not really sure what the best course of action is. The situation is a bit unfortunate. I guess it depends on a few things:
> 
> * Will [Make a beginning of parsing GNU date `items` with nom #25](https://github.com/uutils/parse_datetime/pull/25) be ready soon?
> * How relevant is this PR to [Make a beginning of parsing GNU date `items` with nom #25](https://github.com/uutils/parse_datetime/pull/25), i.e. can we reuse significant parts of this with `nom`?
> 
> The first question is kinda tough. I think a basic `nom` version of this library _should_ be merged soon, just to get the transition going. Which might not be the entirety of #25. The second question is interesting. At the least the tests are relevant after the introduction of `nom`.
> 
> Maybe another approach is to transition to `nom` from the bottom up. What I mean with that is that we port small parts first and gradually build up to parsing the whole thing with `nom`. For example, if we make a `nom`-based` function to parse weekdays, we can easily integrate that later.
> 
> I'm really just explaining my thoughts here. I'm not yet really sure what the best approach is. What do you think?

How about waiting until next week and see the progress of #25. If nothing happens i can rewrite this pr first to use nom and check the pr.

File: src/lib.rs [hunk 1]
```diff
@@ -168,6 +172,24 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
         }
     }
 
+    // parse weekday
+    if let Ok(weekday) = parse_weekday::parse_weekday(s.as_ref()) {
+        let now = Local::now().naive_local();
```
Comment (tertsdiepraam) [inline src/lib.rs:177]: I think we shouldn't use `now` here, but the reference `date`. Also I think we don't need to convert a naive date, which should simplify the code a bit.

File: .gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
 target/
+fuzz/corpus
```
Comment (tertsdiepraam) [inline .gitignore:2]: I think this was fixed in your other PR, but in a `.gitignore` in the `fuzz` directory.

File: src/parse_weekday.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
+use chrono::{ParseWeekdayError, Weekday};
+
+const WEEKDAY_CONVERSION_MAP: [(&str, &str); 3] =
+    [("tues", "tue"), ("thurs", "thu"), ("thus", "thu")];
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/parse_weekday.rs:4]: @tertsdiepraam its "thurs"

**Comment by philolo1** [comment]
@tertsdiepraam thanks for reviewing, but i am currently rewriting it to use nom.

**Comment by philolo1** [comment]
I converted it to draft, sorry.

File: src/parse_weekday.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (philolo1) [reply src/parse_weekday.rs:4]: oh sorry i think i did this by mistake, good catch

**Comment by philolo1** [comment]
@tertsdiepraam it should be ready to review now.

File: src/lib.rs [hunk 4]
```diff
@@ -168,6 +172,24 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
         }
     }
 
+    // parse weekday
+    if let Ok(weekday) = parse_weekday::parse_weekday(s.as_ref()) {
+        let mut beginning_of_day = Local
+            .with_ymd_and_hms(date.year(), date.month(), date.day(), 0, 0, 0)
+            .unwrap();
```
Comment (tertsdiepraam) [inline src/lib.rs:179]: Can't we use `date` directly? Also by using `Local` directly we might get into trouble if we change `date` to not (necessarily) use `Local` as timezone.

File: src/lib.rs [hunk 5]
```diff
@@ -168,6 +172,24 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
...
+
+        loop {
+            if beginning_of_day.weekday() == weekday {
+                break;
+            }
+            beginning_of_day += Duration::days(1);
+        }
```
Comment (tertsdiepraam) [inline src/lib.rs:186]: You've written your own `while` loop :)
```suggestion
        while beginning_of_day.weekday() != weekday {
            beginning_of_day += Duration::days(1);
        }
```

File: src/parse_weekday.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            ("thursday", Thu),
+            ("fri", Fri),
+            ("friday", Fri),
+            ("sat", Sat),
+            ("saturday", Sat),
+            ("sun", Sun),
+            ("sunday", Sun),
+        ];
+
+        for day in days.iter() {
```
Comment (tertsdiepraam) [inline src/parse_weekday.rs:80]: I _think_ `iter` is redundant here.

File: src/parse_weekday.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            test_strings.push(format!("{} ", day.0));
+            test_strings.push(format!(" {}", day.0.to_uppercase()));
+
+            let (left, right) = day.0.split_at(1);
+            test_strings.push(format!("{}{}", left.to_uppercase(), right.to_lowercase()));
+            test_strings.push(format!("{}{}", left.to_lowercase(), right.to_uppercase()));
+
+            for str in test_strings.iter() {
+                assert!(parse_weekday(str).is_ok());
+                assert_eq!(parse_weekday(str).unwrap(), day.1);
```
Comment (tertsdiepraam) [inline src/parse_weekday.rs:94]:
```suggestion
                assert_eq!(parse_weekday(str), Ok(day.1));
```

File: src/parse_weekday.rs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            let mut test_strings = vec![];
+            test_strings.push(String::from(day.0));
+            test_strings.push(format!(" {}", day.0));
+            test_strings.push(format!(" {} ", day.0));
+            test_strings.push(format!("{} ", day.0));
+            test_strings.push(format!(" {}", day.0.to_uppercase()));
+
+            let (left, right) = day.0.split_at(1);
+            test_strings.push(format!("{}{}", left.to_uppercase(), right.to_lowercase()));
+            test_strings.push(format!("{}{}", left.to_lowercase(), right.to_uppercase()));
```
Comment (tertsdiepraam) [inline src/parse_weekday.rs:90]: I'd love it if we can simplify this, because technically this is just a bunch `assert_eq!`s. I think tests are perfectly fine if they are repetitive and should contain as little additional logic as possible.

It could be something like:
```rust
for (string, day) in days {
    assert_eq!(parse_weekday(string), Ok(day));
    assert_eq!(parse_weekday(format!("  {string}"), Ok(day));
    // etc.
}
```

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Interesting! I quite like how that's looking. It's definitely a nice way to start using `nom`!

File: src/parse_weekday.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+use nom::combinator::map;
+use nom::{self, Finish, IResult};
+
+pub(crate) fn parse_weekday(s: &str) -> Result<Weekday, &str> {
+    let s = s.trim();
+
+    let parse_result: IResult<&str, Weekday> = nom::combinator::all_consuming(alt((
+        map(alt((tag_no_case("monday"), tag_no_case("mon"))), |_| {
+            Weekday::Mon
+        }),
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/parse_weekday.rs:13]: @tertsdiepraam I am not sure what "m" means, but it does not seem to be monday.

This is what my ubuntu puts out :
```
ubuntu@ip-172-31-46-7:~$ date -d 'm'
Tue Aug 29 21:00:00 JST 2023
```

**Comment by philolo1** [comment]
@tertsdiepraam i did all the requests accept one, i am not sure what "m" means, but it does not seem to be monday. 

I just tested all letters from a to z with `date` and this is the result:

```
a
Wed Aug 30 08:00:00 JST 2023
b
Wed Aug 30 07:00:00 JST 2023
c
Wed Aug 30 06:00:00 JST 2023
d
Wed Aug 30 05:00:00 JST 2023
e
Wed Aug 30 04:00:00 JST 2023
f
Wed Aug 30 03:00:00 JST 2023
g
Wed Aug 30 02:00:00 JST 2023
h
Wed Aug 30 01:00:00 JST 2023
i
Wed Aug 30 00:00:00 JST 2023
j
date: invalid date ‘j’
k
Tue Aug 29 23:00:00 JST 2023
l
Tue Aug 29 22:00:00 JST 2023
m
Tue Aug 29 21:00:00 JST 2023
n
Wed Aug 30 10:00:00 JST 2023
o
Wed Aug 30 11:00:00 JST 2023
p
Wed Aug 30 12:00:00 JST 2023
q
Wed Aug 30 13:00:00 JST 2023
r
Wed Aug 30 14:00:00 JST 2023
s
Wed Aug 30 15:00:00 JST 2023
t
Wed Aug 30 16:00:00 JST 2023
u
Wed Aug 30 17:00:00 JST 2023
v
Wed Aug 30 18:00:00 JST 2023
w
Wed Aug 30 19:00:00 JST 2023
x
Wed Aug 30 20:00:00 JST 2023
y
Wed Aug 30 21:00:00 JST 2023
z
Wed Aug 30 09:00:00 JST 2023
```

**Comment by philolo1** [comment]
@tertsdiepraam i believe i found the reference source code, apparently "WEDNES" is also valid. https://github.com/coreutils/gnulib/blob/master/lib/parse-datetime.y#L1032

**Comment by philolo1** [comment]
@tertsdiepraam apparently one letter is military time table, i will create another issue. 

https://github.com/coreutils/gnulib/blob/master/lib/parse-datetime.y#L1155

**Comment by philolo1** [comment]
Here is the issue: #39

File: src/parse_weekday.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/parse_weekday.rs]: Ah you're right. It's indeed military time zones. Thanks for figuring that out!

**Review by tertsdiepraam — COMMENTED** [review]
Nice work! I have just two more comments for things that should not be fixed in this PR and one question.

File: src/lib.rs [hunk 10]
```diff
@@ -168,6 +173,27 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
         }
     }
 
+    // parse weekday
+    if let Some(weekday) = parse_weekday::parse_weekday(s.as_ref()) {
+        let mut beginning_of_day = date
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/lib.rs:178]: i dit not find anything either :(

File: src/lib.rs [hunk 11]
```diff
@@ -353,6 +379,56 @@ mod tests {
...
+
+        fn get_formatted_date(date: DateTime<Local>, weekday: &str) -> String {
+            let result = parse_datetime_at_date(date, weekday).unwrap();
+
+            return result.format("%F %T %f").to_string();
+        }
+        #[test]
+        fn test_weekday() {
+            // add some constant hours and minutes and seconds to check its reset
+            let date = Local.with_ymd_and_hms(2023, 02, 28, 10, 12, 3).unwrap();
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/lib.rs:396]: I am not sure how to best not use local but agreed.

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/lib.rs:396]: Yeah, `parse_datetime_at_date` should just be generic over the timezone. And then it can be called with `Utc`, `Local` or `FixedOffset`. Like I said though, definitely out of scope for this PR.

File: src/lib.rs [hunk 12]
```diff
@@ -168,6 +173,27 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
...
+            .with_second(0)
+            .unwrap()
+            .with_nanosecond(0)
+            .unwrap();
+
+        while beginning_of_day.weekday() != weekday {
+            beginning_of_day += Duration::days(1);
+        }
+
+        if let Ok(dt) = naive_dt_to_fixed_offset(date, beginning_of_day.naive_local()) {
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/lib.rs]: @tertsdiepraam seems it works, thanks.

**Comment by philolo1** [comment]
@tertsdiepraam Thank for the review. I am trying to improve my rust coding skills and contributing to open source is a good opportunity.

**Comment by philolo1** [comment]
@tertsdiepraam I am not to familiar with github, it still shows changed requested. Do i have to do something to request review. Furthermore, should i squash the commits or would you do that when you merge it?

**Comment by philolo1** [comment]
I squashed, would be nice if someone can get this merged.

File: src/parse_weekday.rs [hunk 13]
```diff
@@ -0,0 +1,96 @@
+use chrono::Weekday;
```
Comment (sylvestre) [inline src/parse_weekday.rs:1]: please add the license header

File: src/parse_weekday.rs [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            alt((tag("thursday"), tag("thurs"), tag("thur"), tag("thu"))),
+        ),
+        value(Weekday::Fri, alt((tag("friday"), tag("fri")))),
+        value(Weekday::Sat, alt((tag("saturday"), tag("sat")))),
+        value(Weekday::Sun, alt((tag("sunday"), tag("sun")))),
+    )))(s);
+
+    match parse_result {
+        Ok((_, weekday)) => Some(weekday),
+        Err(_) => None,
```
Comment (sylvestre) [inline src/parse_weekday.rs:29]: why don't you return a Result ?
here, it would be 
        Err(_) => Err("Failed to parse weekday"),

Comment (philolo1) [reply src/parse_weekday.rs:29]: I had this discussion with @tertsdiepraam, but I think the reason is this is used in `parse_datetime` and we only care about whether it was parsed or not.

File: src/parse_weekday.rs (hunk 13 above)
Comment (philolo1) [reply src/parse_weekday.rs:3]: done

File: src/parse_weekday.rs (hunk 14 above)
Comment (tertsdiepraam) [reply src/parse_weekday.rs:31]: My reasoning was that we only care about the parsing the whole datetime. If a weekday fails, another type of item might still succeedsl.

Comment (philolo1) [reply src/parse_weekday.rs:31]: @sylvestre should i change it to Result

Comment (philolo1) [reply src/parse_weekday.rs:31]: @tertsdiepraam @sylvestre its been a while, i am fine with whatever but it would be good to get this pr also merged, before we forget what it was about :) .

**Review by sylvestre — COMMENTED** [review]
sorry, i didn't submitted my comments :(

File: src/parse_weekday.rs [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+use nom::bytes::complete::tag;
+use nom::combinator::value;
+use nom::{self, IResult};
+
+pub(crate) fn parse_weekday(s: &str) -> Option<Weekday> {
+    let s = s.trim().to_lowercase();
+    let s = s.as_str();
+
+    let parse_result: IResult<&str, Weekday> = nom::combinator::all_consuming(alt((
+        value(Weekday::Mon, alt((tag("monday"), tag("mon")))),
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/parse_weekday.rs]: @sylvestre done

**Comment by sylvestre** [comment]
thanks
